shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -491,8 +495,12 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
             val reverseRigid2 = reverseRigid ++ dummyTargs2.map(t =>
               t -> freshVar(t.prov, S(t.id.idStr))(thisCtx.lvl + 1))
             val bodyTy = subst(rhs.fold(
-              term =>
-                typeLetRhs(rec, nme.name, term)(thisCtx, raise, targsMap2),
+              term => ctx.getMthDefn(prt.name, nme.name, false)
+                .fold(typeLetRhs(rec, nme.name, term)(thisCtx, raise, targsMap2))(_.toPT match {
+                  case PolymorphicType(level, ProvType(underlying)) =>
```

Comment:
This look pretty weird and ad-hoc. I'd like a comment explaining why it's necessary.

Reply:
It's not *that* necessary, but it makes the error reporting nicer. I have added a comment to explain that.